### DISCUSSION TITLE
Getting cookies value on App init

### DISCUSSION THREAD

**Discussion by gsambrotta** [vercel/next.js#20394] [Help]
Hi there, 
In my app I need to get the cookies value on the first render. 
I get those in my `_app.tsx`. Everything looks fine (render correctly the html) but I get the server/client mismatch warning because at the first render on Server, cookies are undefined and the value fall back to default value which is 0.
On hydration, the value is picked from cookies and is displayed correctly.

Could someone explain me why is a problem that on the server the value is the default value (therefor why I get this warning) and what would be a better way to write this code?

Here my `_app.tsx`
```
import React, { useState, useEffect } from 'react'
import type { AppProps } from 'next/app'
import { UserContext } from '../context/UserContext'
require('es6-promise').polyfill()

let cookieHelper
if (typeof window !== 'undefined') {
  cookieHelper = require( '../helpers/_cookies' ) // This is a file written by us where we export const and get/set cookies func 
}

function ElliotApp ({ Component, pageProps }: AppProps) {
  useEffect(() => {
    import('../helpers/_cookies')
  }, [])
  
  const searchesFromCookies = cookieHelper?.get(cookieHelper?.COOKIE_NAME_SEARCH_COUNT) // this value is a string like '3'
  const userState = { 
    numOfSearches: searchesFromCookies || 0 
  }
  const [userContext, setUserContext] = useState(userState)

  useEffect(() => {
    cookieHelper?.set(cookieHelper?.COOKIE_NAME_SEARCH_COUNT, userContext.numOfSearches)
  }, [userContext])

  return (
    <UserContext.Provider value={[userContext, setUserContext]}>
      <Component {...pageProps} />
    </UserContext.Provider>
  )
}

export default ElliotApp

```

**Comment by jamesmosier** [comment]
The server is not able to get the cookies by using the browser method you are using to obtain them. Instead you could need to get the cookies from the request on the server. If you wanted to get the cookies on the server inside `_app.tsx` you would need to use `getInitialProps` and return the value(s) you wanted as props. Keep in mind by doing this you would then be server rendering every page in your application.

```js
// Only use this method if you have blocking data requirements for
// every single page in your application. This disables the ability to
// perform automatic static optimization, causing every page in your app to
// be server-side rendered.
//
ElliotApp.getInitialProps = async (appContext) => {
  // calls page's `getInitialProps` and fills `appProps.pageProps`
  const appProps = await App.getInitialProps(appContext);
  // you will need to parse the cookies by writing a method yourself or using something like the `cookie` npm package
  const cookies = parseTheCookies(appContext.req.headers.cookie)
  const searchesFromCookie = cookies[COOKIE_NAME_SEARCH_COUNT]
  return { ...appProps, searchesFromCookie }
}
```

[Docs](https://nextjs.org/docs/advanced-features/custom-app)


---

On an unrelated note, you do not need to use `require('es6-promise').polyfill()` as Next polyfills APIs like this by default for older browsers.

**Reply by gsambrotta** [reply]
Hey @jamesmosier thank you for your answer! 
yes, I'm aware that I cannot get cookies in the server with the browser method, but that's my point. I don't need the cookies value on the server so I'm fine with having a default value on the sever side and then update it on the client. 
And it does work.. but I get the mismatch warning. And I don't get why is a problem.

For my understanding `getInitialProps` is deprecated (or it will soon) and when I used for other pages I got an annoying warning in the console suggesting me to switch to `getServerSideProps`.
Also, I'm not sure I want all my pages to be SSR. It will SSR each page because I'm using this method in the `_app.js` Therefor all the pages under it will use the same method and I cannot differentate anymore. Correct?

Thanks for the es6-promise polyfill tip! :D
